### DISCUSSION TITLE
정말 너무 쉬운 Docker - peterkimzz

### DISCUSSION THREAD

**Discussion by giscus** [peterkimzz/blog#18] [Blog Comments]
# 정말 너무 쉬운 Docker - peterkimzz

우리가 Docker를 사용해야하는 가장 큰 이유는, 어떤 컴퓨터에서든 똑같은 개발 환경을 보장해주기 떄문입니다.
로컬 컴퓨터에서 열심히 개발하고 AWS에 코드를 올렸는데, 에러를 마주하며 스트레스를 받았던 경험이 한 번쯤은 있을겁니다. 내 컴퓨터랑 클라우드 컴퓨터의 환

https://www.peterkimzz.com/super-easy-docker/

**Comment by giwonn** [comment]
몇가지 궁금한 점이 있어 댓글 남깁니다!  

---  

Dockerfile에서 `COPY . .`를 하게 되면 전부 복사가 되는거니까 전부 복사하고 yarn install을 실행해도 될것같은데요.  
package.json, yarn.lock을 먼저 카피하여 install 하신 이유가 궁금합니다!  
게다가 `COPY . .'으로 인해서 node_modules 또한 복사될것같은데... 이렇게 생각하면 install도 필요없을것 같아서요. 헷갈리네요 ㅠㅠ  

---  

docker compose에서 volumes에 `app/node_modules`를 추가하셨는데요. 이것으로 핫 리로딩이 개선된다는게 무슨 뜻인지 잘 모르겠습니다!  
이것 또한  `.:/usr/app/api`로 인해서 node_modules가 관리될것 같은데.. 잘 모르겠네요 ㅠ

**Reply by peterkimzz** [reply]
1.

도커의 가장 큰 장점 중 하나는 **캐싱** 입니다. 이미지의 전과 후를 비교해 변화가 없는 부분은 미리 빌드된 이미지를 이용해 빠르게 다시 이미지를 빌드해준다는 점이죠.

하지만 말씀하신대로 `package.json` 만 복사하는 과정 없이 전체를 `COPY . .` 하게되면 캐싱의 이점을 누릴 수 없게됩니다. 이미지를 빌드할 때마다 용량이 매우 큰 `node_modules` 폴더를 포함해서 프로젝트 소스코드 전체를 복사, 붙여넣기 하는 과정이 들어가기 때문이죠. 이 부분은 실제로 테스트 해보시면 꽤 큰 차이가 납니다.

2. 

이거는 저도 100% 이해하고 있지는 않은데, 도커의 볼륨 마운트 동작 방식과 관련이 있는걸로 알고 있습니다.

저의 경우엔 `.:/usr/app/api` 만 넣었을 땐 핫 리로드가 되지 않았고, `node_modules` 만 따로 지정해서 볼륨을 마운트해주었을 땐 핫 리로드가 적용이 되었습니다. 아마 호스트 (내 컴퓨터)의 `node_modules` 에는 패키지가 설치가 되지 않는 문제를 해결할 수 있는 것으로 알고 있는데, 질문 주신것을 보고 저도 구글링을 조금 해봤는데 아래 링크를 참고해주시면 좀 더 도움이 되실 것 같습니다. 

https://ui.toast.com/weekly-pick/ko_20160823

질문주셔서 감사합니다. 저도 요새는 도커를 잘 활용하지 않아서 기억이 가물가물한데, 쉬는 날에 테스트 해보고 글 내용 업데이트를 해야겠네요~!

**Reply by giwonn** [reply]
자세한 답변 정말 감사드립니다.  
  
덕분에 이 부분에 대해서는 확실하게 이해한것 같아요!
